tests/30rooms/12thirdpartyinvite.pl

```diff
@@ -0,0 +1,395 @@
+#  4. Asserts that invitee did/didn't join the room, depending on truthiness of expect_join_success
+#  5. Awaits on all passed futures, so that you can stub/mock things as you wish
+sub make_3pid_invite {
+   my %args = @_;
+   my $inviter = $args{inviter};
+   my $invitee = $args{invitee};
+   my $expect_join_success = $args{expect_join_success};
+   my $join_sub = $args{join_sub};
+   my $is_user_agent = $args{is_user_agent};
+
+   my $invitee_email = 'lemurs@monkeyworld.org';
+   my $token = "abc123";
+
+   my $response_verifier = $expect_join_success
+      ? sub { $_[0] } : \&main::expect_http_4xx;
+
+
+   my $stub_id_server = SyTest::Identity::Server->new;
+   $loop->add( $stub_id_server );
+   $stub_id_server->listen(
+      host    => "localhost",
+      service => "",
+      extensions => [qw( SSL )],
+      # Synapse currently only talks IPv4
+      family => "inet",
+
+      SSL_cert_file => "$DIR/../../keys/tls-selfsigned.crt",
+      SSL_key_file => "$DIR/../../keys/tls-selfsigned.key",
+   )->then( sub {
+      my ( $listener ) = @_;
+      my $sock = $listener->read_handle;
+      my $id_server = sprintf "%s:%d", $sock->sockhostname, $sock->sockport;
+      if( defined $is_user_agent ) {
+         $stub_id_server->{isvalid_needs_useragent} = $is_user_agent;
+      }
+
+      my @is_valid_stubs;
+      my $room_id;
+
+      Future->needs_all(
+         matrix_create_room( $inviter, visibility => "private" )
+         ->then(sub {
+            ( $room_id ) = @_;
+            $stub_id_server->stub_token( $token, "email", $invitee_email, $inviter->user_id, $room_id );
+            do_3pid_invite( $inviter, $room_id, $id_server, $invitee_email )
+         })->then( sub {
+            my $signature = encode_base64_unpadded( $crypto_sign->mac( $token, $stub_id_server->{private_key} ) );
+            my $signatures = { $id_server => { "ed25519:0" => $signature } };
+            $join_sub->( $token, $stub_id_server->{keys}{"ed25519:0"}, $signatures, $room_id, $id_server, $stub_id_server )
+         })->followed_by($response_verifier)
+         ->then( sub {
+            matrix_get_room_state( $inviter, $room_id,
+               type      => "m.room.member",
+               state_key => $invitee->user_id,
+            )
+         })->followed_by(assert_membership( $inviter, $expect_join_success ? "join" : undef ) ),
+      );
+   });
+}
+
+sub assert_membership {
+   my ( $user, $expected_membership ) = @_;
+
+   my $verifier = defined $expected_membership
+      ? sub {
+         my ( $f ) = @_;
+
+         $f->then( sub {
+            my ( $body ) = @_;
+
+            log_if_fail "Body", $body;
+            $body->{membership} eq $expected_membership or
+               die "Expected invited user membership to be '$expected_membership'";
+
+            Future->done( 1 );
+         } )
+      }
+      : \&main::expect_http_error;
+};
+
+sub do_3pid_invite {
```

Comment:
Is the process of sending 3PID invites a thing that's likely to be useful in other test files too? If so, consider naming this `matrix_...` and using the `@EXPORT` mechanism to make it available to later files.